soroban-env-host/src/host/frame.rs

```diff
@@ -196,7 +207,17 @@ impl Host {
             // See explanation for this line in [crate::vm::Vm::parse_module] -- it exists
             // to add-back module-parsing costs that were suppressed during the invocation.
             if self.in_storage_recording_mode()? {
-                self.rebuild_module_cache()?;
+                let is_lifecycle_function_frame = if let Some(c) = &ctx {
+                    c.frame.is_lifecycle_host_function_frame()
+                } else {
+                    false
+                };
+                if !is_lifecycle_function_frame {
+                    // Host function calls that upload Wasm and create contracts
+                    // don't use the module cache and thus don't need to have it
+                    // rebuilt.
```

Comment:
```suggestion
                    // rebuilt.
                    //
                    // Note: VM-running contracts that call `upload_wasm` _will_ get here, because they did not
                    // push a lifecycle function frame, but that's _nearly_ correct: such a contract does get a
                    // module cache (containing the running contract), it just doesn't get one containing the
                    // uploaded contract. So we'll be over-estimating cost a bit in that case by simulating a
                    // module-cache build including both contracts, instead of just the running contract.
```

Reply:
Done